### DISCUSSION TITLE
Adding Heroku's buildpacks as a builder option

### DISCUSSION THREAD

**Discussion by nickhammond** [basecamp/kamal#795] [Ideas]
Heroku recently open-sourced their [Cloud Native Buildpack](https://github.com/heroku/buildpacks) and it's worth exploring as an alternative builder option. This project is essentially the Heroku build process, a very familiar build process that just figures out a bunch of the basic stuff for you by checking for various files within your app.

Docker often comes up as a stumbling block on Discord and GitHub discussions for Kamal as people transition from other PAAS offerings or perhaps provisioned VPS machines. There's also a lot of reinventing the wheel for things like caching asset compilation or bundle install caching with Docker builds. 

The Heroku buildpacks have been doing this for a long time and it's also a shared effort to get a Ruby, Node, Python, etc. app hosted. There are [9 officially supported buildpacks](https://elements.heroku.com/buildpacks) from Heroku but another [7800 from the community](https://elements.heroku.com/buildpacks) that you can leverage with Heroku buildpacks. The buildpacks are also Cloud Native Buildpacks which is now an [open spec by the Cloud Native Computing Foundation](https://github.com/heroku/buildpacks?tab=readme-ov-file#what-is-a-cloud-native-buildpack) (CNCF) and no longer bespoke to Heroku.

Currently with Kamal we can almost accomplish this with a `pre-build` script that packs our app and then references the packed image in our `deploy.yml` that we just pushed up to our registry. Something like this:

``` bash
pack build hey
docker push hey
```
_We'd want to also tag the container with the `KAMAL_VERSION` with these steps._

Packing our app results in a fully built OCI-compliant container image that we can then push up to our registry host. For example, here's a build for a Rails app with the `heroku-22` buildpack:

<details>
<summary>First build output</summary>

```
time pack build test-heroku
22: Pulling from heroku/builder
Digest: sha256:ba1e74c72257218a22d72049f6df3c60ae02a43a940cde9ad80e0435145cf7d9
Status: Image is up to date for heroku/builder:22
22-cnb: Pulling from heroku/heroku
Digest: sha256:b1dedd0404b81a545d55739bf86ff95757d3a0544d272ada9f67e0eaf2b10b11
Status: Image is up to date for heroku/heroku:22-cnb
===> ANALYZING
Restoring data for SBOM from previous image
===> DETECTING
4 of 5 buildpacks participating
heroku/nodejs-engine 3.0.6
heroku/nodejs-yarn   3.0.6
heroku/ruby          2.1.3
heroku/procfile      3.1.1
===> RESTORING
Restoring metadata for "heroku/ruby:bundler" from app image
Restoring metadata for "heroku/ruby:cache_public_assets" from app image
Restoring metadata for "heroku/ruby:cache_tmp_cache_assets" from app image
Restoring metadata for "heroku/ruby:gems" from app image
Restoring metadata for "heroku/ruby:ruby" from app image
Restoring data for "heroku/ruby:bundler" from cache
Restoring data for "heroku/ruby:cache_public_assets" from cache
Restoring data for "heroku/ruby:cache_tmp_cache_assets" from cache
Restoring data for "heroku/ruby:gems" from cache
Restoring data for "heroku/ruby:ruby" from cache
===> BUILDING

[Heroku Node.js Engine Buildpack]

[Checking Node.js version]
Node.js version not specified, using 20.x
Resolved Node.js version: 20.12.2

[Installing Node.js distribution]
Downloading Node.js 20.12.2
Extracting Node.js 20.12.2
Installing Node.js 20.12.2

[Detecting yarn CLI version to install]
No yarn engine range detected in package.json, using default ({DEFAULT_YARN_REQUIREMENT})
Resolved yarn CLI version: 1.22.22

[Installing yarn CLI]
Downloading yarn 1.22.22
Extracting yarn 1.22.22
Installing yarn 1.22.22
Yarn CLI operating in yarn 1.22.22 mode.

[Setting up yarn dependency cache]
yarn config v1.22.22
success Set "cache-folder" to "/layers/heroku_nodejs-yarn/deps/cache".
Done in 0.04s.

[Installing dependencies]
yarn install v1.22.22
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
Done in 1.94s.

[Running scripts]
No build scripts found

# Heroku Ruby Buildpack

- Metrics agent
  - Skipping install (`barnes` gem not found)
- Ruby version `3.2.2` from `Gemfile.lock`
  - Clearing cache (ruby version changed)
  - Installing ...... (3.281s)
- Bundler version `2.4.12` from `Gemfile.lock`
  - Clearing cache (bundler version changed)
  - Running `gem install bundler --version 2.4.12` ... (0.840s)
- Bundle install
  - Clearing cache (ruby version changed)
  - Running `BUNDLE_BIN="/layers/heroku_ruby/gems/bin" BUNDLE_CLEAN="1" BUNDLE_DEPLOYMENT="1" BUNDLE_GEMFILE="/workspace/Gemfile" BUNDLE_PATH="/layers/heroku_ruby/gems" BUNDLE_WITHOUT="development:test" bundle install`

      Fetching gem metadata from https://rubygems.org/.........
      Fetching rake 13.1.0
      Installing rake 13.1.0
      Fetching base64 0.2.0
      Fetching bigdecimal 3.1.5
      Fetching concurrent-ruby 1.2.3
      Fetching connection_pool 2.4.1
      Using ruby2_keywords 0.0.5
      Fetching minitest 5.20.0
      Fetching mutex_m 0.2.0
      Fetching builder 3.2.4
      Fetching erubi 1.12.0
      Fetching racc 1.7.3
      Fetching rack 3.0.9.1
      Fetching crass 1.0.6
      Fetching nio4r 2.7.0
      Installing base64 0.2.0
      Installing crass 1.0.6
      Fetching websocket-extensions 0.1.5
      Installing mutex_m 0.2.0
      Installing erubi 1.12.0
      Installing builder 3.2.4
      Installing connection_pool 2.4.1
      Fetching zeitwerk 2.6.12
      Fetching timeout 0.4.1
      Fetching marcel 1.0.2
      Fetching mini_mime 1.1.5
      Installing bigdecimal 3.1.5 with native extensions
      Fetching date 3.3.4
      Installing racc 1.7.3 with native extensions
      Installing websocket-extensions 0.1.5
      Installing timeout 0.4.1
      Fetching msgpack 1.7.2
      Installing minitest 5.20.0
      Installing nio4r 2.7.0 with native extensions
      Using bundler 2.4.12
      Fetching stringio 3.1.0
      Installing zeitwerk 2.6.12
      Installing mini_mime 1.1.5
      Installing marcel 1.0.2
      Installing rack 3.0.9.1
      Installing date 3.3.4 with native extensions
      Fetching io-console 0.7.1
      Fetching webrick 1.8.1
      Fetching thor 1.3.0
      Installing concurrent-ruby 1.2.3
      Fetching dotenv 2.8.1
      Installing stringio 3.1.0 with native extensions
      Installing msgpack 1.7.2 with native extensions
      Installing io-console 0.7.1 with native extensions
      Installing dotenv 2.8.1
      Installing thor 1.3.0
      Installing webrick 1.8.1
      Fetching hanami-utils 0.9.2
      Fetching sqlite3 1.7.0 (x86_64-linux)
      Fetching tilt 2.3.0
      Fetching drb 2.2.0
      Installing hanami-utils 0.9.2
      Installing tilt 2.3.0
      Fetching redis-client 0.19.1
      Fetching websocket-driver 0.7.6
      Fetching net-protocol 0.2.2
      Installing drb 2.2.0
      Fetching rack-session 2.0.0
      Installing net-protocol 0.2.2
      Installing redis-client 0.19.1
      Installing websocket-driver 0.7.6 with native extensions
      Installing rack-session 2.0.0
      Fetching rack-test 2.1.0
      Fetching url_mount 0.2.1
      Fetching rackup 2.1.0
      Installing rack-test 2.1.0
      Installing rackup 2.1.0
      Fetching i18n 1.14.1
      Fetching tzinfo 2.0.6
      Installing url_mount 0.2.1
      Installing sqlite3 1.7.0 (x86_64-linux)
      Fetching redis 5.0.8
      Installing i18n 1.14.1
      Installing redis 5.0.8
      Fetching net-pop 0.1.2
      Installing tzinfo 2.0.6
      Fetching net-smtp 0.4.0.1
      Fetching http_router 0.11.2
      Installing net-smtp 0.4.0.1
      Installing net-pop 0.1.2
      Installing http_router 0.11.2
      Fetching hanami-router 0.6.2
      Installing hanami-router 0.6.2
      Fetching nokogiri 1.16.0 (x86_64-linux)
      Fetching psych 5.1.2
      Installing psych 5.1.2 with native extensions
      Installing nokogiri 1.16.0 (x86_64-linux)
      Fetching loofah 2.22.0
      Installing loofah 2.22.0
      Fetching rails-html-sanitizer 1.6.0
      Installing rails-html-sanitizer 1.6.0
      Fetching rdoc 6.6.2
      Installing rdoc 6.6.2
      Fetching reline 0.4.2
      Installing reline 0.4.2
      Fetching irb 1.11.0
      Installing irb 1.11.0
      Fetching puma 6.4.2
      Installing puma 6.4.2 with native extensions
      Fetching net-imap 0.4.9.1
      Installing net-imap 0.4.9.1
      Fetching mail 2.8.1
      Installing mail 2.8.1
      Fetching bootsnap 1.17.0
      Installing bootsnap 1.17.0 with native extensions
      Fetching activesupport 7.1.2
      Fetching oj 3.16.3
      Installing oj 3.16.3 with native extensions
      Installing activesupport 7.1.2
      Fetching activemodel 7.1.2
      Fetching globalid 1.2.1
      Fetching rails-dom-testing 2.2.0
      Installing rails-dom-testing 2.2.0
      Installing globalid 1.2.1
      Fetching activejob 7.1.2
      Fetching actionview 7.1.2
      Installing activemodel 7.1.2
      Fetching activerecord 7.1.2
      Installing activejob 7.1.2
      Installing actionview 7.1.2
      Fetching actionpack 7.1.2
      Fetching jbuilder 2.11.5
      Installing activerecord 7.1.2
      Installing jbuilder 2.11.5
      Installing actionpack 7.1.2
      Fetching actioncable 7.1.2
      Fetching activestorage 7.1.2
      Fetching railties 7.1.2
      Fetching actionmailer 7.1.2
      Installing actionmailer 7.1.2
      Installing actioncable 7.1.2
      Installing activestorage 7.1.2
      Fetching actionmailbox 7.1.2
      Fetching actiontext 7.1.2
      Installing railties 7.1.2
      Fetching dotenv-rails 2.8.1
      Fetching importmap-rails 2.0.1
      Fetching propshaft 0.8.0
      Fetching cssbundling-rails 1.4.0
      Fetching stimulus-rails 1.3.3
      Fetching turbo-rails 1.5.0
      Installing actionmailbox 7.1.2
      Installing actiontext 7.1.2
      Fetching rails 7.1.2
      Installing cssbundling-rails 1.4.0
      Installing importmap-rails 2.0.1
      Installing propshaft 0.8.0
      Installing dotenv-rails 2.8.1
      Installing stimulus-rails 1.3.3
      Installing turbo-rails 1.5.0
      Installing rails 7.1.2
      Fetching heroicon 1.0.0
      Installing heroicon 1.0.0
      Fetching litestack 0.4.2
      Installing litestack 0.4.2
      Bundle complete! 28 Gemfile dependencies, 81 gems now installed.
      Gems in the groups 'development' and 'test' were not installed.
      Bundled gems are installed into `/layers/heroku_ruby/gems`

  - Done (54.113s)
- Setting default processes
  - Running `bundle list` ... (0.244s)
  - Detected rails app (`rails` gem found)
- Rake assets install
  - Detected rake (`rake` gem found, `Rakefile` found at `/workspace/Rakefile`)
  - Running `bundle exec rake -P --trace` ... (0.912s)
  - Compiling assets with cache (detected `rake assets:precompile` and `rake assets:clean` via `bundle exec rake -P`)
  - Loading cache for /workspace/public/assets
  - Loading cache for /workspace/tmp/cache/assets
  - Running `bundle exec rake assets:precompile assets:clean --trace`

      ** Invoke assets:precompile (first_time)
      ** Invoke environment (first_time)
      ** Execute environment
      ** Invoke css:build (first_time)
      ** Invoke css:install (first_time)
      ** Execute css:install
      yarn install v1.22.22
      [1/4] Resolving packages...
      success Already up-to-date.
      Done in 0.09s.
      ** Execute css:build
      yarn run v1.22.22
      $ sass ./app/assets/stylesheets/application.sass.scss:./app/assets/builds/application.css --no-source-map --load-path=node_modules
      Done in 0.53s.
      ** Execute assets:precompile
      I, [2024-05-09T12:46:43.008018 #1176]  INFO -- : Writing application-c94f371f614ee5c699e4d83ccb293caa7a53e540.css
      I, [2024-05-09T12:46:43.083975 #1176]  INFO -- : Writing application-bfcdf840723fd1f2a658705badd7942656d5f087.js
  - Done (2.546s)
  - Storing cache for /workspace/public/assets
  - Storing cache for /workspace/tmp/cache/assets
- Done (finished in 1m 2s)

[Discovering process types]
Procfile declares types -> web
===> EXPORTING
Adding layer 'heroku/nodejs-engine:dist'
Adding layer 'heroku/nodejs-engine:node_runtime_metrics'
Adding layer 'heroku/nodejs-engine:web_env'
Adding layer 'heroku/nodejs-yarn:deps'
Adding layer 'heroku/nodejs-yarn:dist'
Adding layer 'heroku/ruby:bundler'
Adding layer 'heroku/ruby:cache_public_assets'
Reusing layer 'heroku/ruby:cache_tmp_cache_assets'
Reusing layer 'heroku/ruby:env_defaults'
Adding layer 'heroku/ruby:gems'
Adding layer 'heroku/ruby:ruby'
Reusing layer 'buildpacksio/lifecycle:launch.sbom'
Adding 1/1 app layer(s)
Reusing layer 'buildpacksio/lifecycle:launcher'
Adding layer 'buildpacksio/lifecycle:config'
Reusing layer 'buildpacksio/lifecycle:process-types'
Adding label 'io.buildpacks.lifecycle.metadata'
Adding label 'io.buildpacks.build.metadata'
Adding label 'io.buildpacks.project.metadata'
Setting default process type 'web'
Saving test-heroku...
*** Images (08111c97fcf2):
      test-heroku
Adding cache layer 'heroku/nodejs-engine:dist'
Adding cache layer 'heroku/nodejs-yarn:deps'
Adding cache layer 'heroku/nodejs-yarn:dist'
Adding cache layer 'heroku/ruby:bundler'
Adding cache layer 'heroku/ruby:cache_public_assets'
Reusing cache layer 'heroku/ruby:cache_tmp_cache_assets'
Adding cache layer 'heroku/ruby:gems'
Adding cache layer 'heroku/ruby:ruby'
Successfully built image test-heroku
pack build test-heroku  0.13s user 0.11s system 0% cpu 1:35.84 total
```
</details>

<details>
<summary>Second build output</summary>

```
➜ time pack build test-heroku
22: Pulling from heroku/builder
Digest: sha256:ba1e74c72257218a22d72049f6df3c60ae02a43a940cde9ad80e0435145cf7d9
Status: Image is up to date for heroku/builder:22
22-cnb: Pulling from heroku/heroku
Digest: sha256:b1dedd0404b81a545d55739bf86ff95757d3a0544d272ada9f67e0eaf2b10b11
Status: Image is up to date for heroku/heroku:22-cnb
===> ANALYZING
Restoring data for SBOM from previous image
===> DETECTING
4 of 5 buildpacks participating
heroku/nodejs-engine 3.0.6
heroku/nodejs-yarn   3.0.6
heroku/ruby          2.1.3
heroku/procfile      3.1.1
===> RESTORING
Restoring metadata for "heroku/nodejs-engine:dist" from app image
Restoring metadata for "heroku/nodejs-engine:node_runtime_metrics" from app image
Restoring metadata for "heroku/nodejs-engine:web_env" from app image
Restoring metadata for "heroku/nodejs-yarn:deps" from app image
Restoring metadata for "heroku/nodejs-yarn:dist" from app image
Restoring metadata for "heroku/ruby:bundler" from app image
Restoring metadata for "heroku/ruby:cache_public_assets" from app image
Restoring metadata for "heroku/ruby:cache_tmp_cache_assets" from app image
Restoring metadata for "heroku/ruby:gems" from app image
Restoring metadata for "heroku/ruby:ruby" from app image
Restoring data for "heroku/nodejs-engine:dist" from cache
Restoring data for "heroku/nodejs-yarn:deps" from cache
Restoring data for "heroku/nodejs-yarn:dist" from cache
Restoring data for "heroku/ruby:bundler" from cache
Restoring data for "heroku/ruby:cache_public_assets" from cache
Restoring data for "heroku/ruby:cache_tmp_cache_assets" from cache
Restoring data for "heroku/ruby:gems" from cache
Restoring data for "heroku/ruby:ruby" from cache
===> BUILDING

[Heroku Node.js Engine Buildpack]

[Checking Node.js version]
Node.js version not specified, using 20.x
Resolved Node.js version: 20.12.2

[Installing Node.js distribution]
Reusing Node.js 20.12.2

[Detecting yarn CLI version to install]
No yarn engine range detected in package.json, using default ({DEFAULT_YARN_REQUIREMENT})
Resolved yarn CLI version: 1.22.22

[Installing yarn CLI]
Reusing yarn 1.22.22
Yarn CLI operating in yarn 1.22.22 mode.

[Setting up yarn dependency cache]
Restoring yarn dependency cache
yarn config v1.22.22
success Set "cache-folder" to "/layers/heroku_nodejs-yarn/deps/cache".
Done in 0.05s.

[Installing dependencies]
yarn install v1.22.22
[1/4] Resolving packages...
[2/4] Fetching packages...
[3/4] Linking dependencies...
[4/4] Building fresh packages...
Done in 0.23s.

[Running scripts]
No build scripts found

# Heroku Ruby Buildpack

- Metrics agent
  - Skipping install (`barnes` gem not found)
- Ruby version `3.2.2` from `Gemfile.lock`
  - Using cached version
- Bundler version `2.4.12` from `Gemfile.lock`
  - Using cached version
- Bundle install
  - Loading cache
  - Skipping `bundle install` (no changes found in /workspace/Gemfile, /workspace/Gemfile.lock, or user configured environment variables)
  - ! HELP To force run `bundle install` set `HEROKU_SKIP_BUNDLE_DIGEST=1`
- Setting default processes
  - Running `bundle list` ... (0.262s)
  - Detected rails app (`rails` gem found)
- Rake assets install
  - Detected rake (`rake` gem found, `Rakefile` found at `/workspace/Rakefile`)
  - Running `bundle exec rake -P --trace` ... (0.967s)
  - Compiling assets with cache (detected `rake assets:precompile` and `rake assets:clean` via `bundle exec rake -P`)
  - Loading cache for /workspace/public/assets
  - Loading cache for /workspace/tmp/cache/assets
  - Running `bundle exec rake assets:precompile assets:clean --trace`

      ** Invoke assets:precompile (first_time)
      ** Invoke environment (first_time)
      ** Execute environment
      ** Invoke css:build (first_time)
      ** Invoke css:install (first_time)
      ** Execute css:install
      yarn install v1.22.22
      [1/4] Resolving packages...
      success Already up-to-date.
      Done in 0.10s.
      ** Execute css:build
      yarn run v1.22.22
      $ sass ./app/assets/stylesheets/application.sass.scss:./app/assets/builds/application.css --no-source-map --load-path=node_modules
      Done in 0.52s.
      ** Execute assets:precompile
      ** Invoke assets:clean (first_time)
      ** Invoke environment
      ** Execute assets:clean

  - Done (2.580s)
  - Storing cache for /workspace/public/assets
  - Storing cache for /workspace/tmp/cache/assets
- Done (finished in 3.992s)

[Discovering process types]
Procfile declares types -> web
===> EXPORTING
Reusing layer 'heroku/nodejs-engine:dist'
Reusing layer 'heroku/nodejs-engine:node_runtime_metrics'
Reusing layer 'heroku/nodejs-engine:web_env'
Reusing layer 'heroku/nodejs-yarn:deps'
Reusing layer 'heroku/nodejs-yarn:dist'
Reusing layer 'heroku/ruby:bundler'
Reusing layer 'heroku/ruby:cache_public_assets'
Reusing layer 'heroku/ruby:cache_tmp_cache_assets'
Reusing layer 'heroku/ruby:env_defaults'
Reusing layer 'heroku/ruby:gems'
Reusing layer 'heroku/ruby:ruby'
Reusing layer 'buildpacksio/lifecycle:launch.sbom'
Adding 1/1 app layer(s)
Reusing layer 'buildpacksio/lifecycle:launcher'
Reusing layer 'buildpacksio/lifecycle:config'
Reusing layer 'buildpacksio/lifecycle:process-types'
Adding label 'io.buildpacks.lifecycle.metadata'
Adding label 'io.buildpacks.build.metadata'
Adding label 'io.buildpacks.project.metadata'
Setting default process type 'web'
Saving test-heroku...
*** Images (4f068fb49dde):
      test-heroku
Reusing cache layer 'heroku/nodejs-engine:dist'
Reusing cache layer 'heroku/nodejs-yarn:deps'
Reusing cache layer 'heroku/nodejs-yarn:dist'
Reusing cache layer 'heroku/ruby:bundler'
Reusing cache layer 'heroku/ruby:cache_public_assets'
Reusing cache layer 'heroku/ruby:cache_tmp_cache_assets'
Reusing cache layer 'heroku/ruby:gems'
Reusing cache layer 'heroku/ruby:ruby'
Successfully built image test-heroku
pack build test-heroku  0.12s user 0.10s system 1% cpu 14.822 total
```
</details>

The first pack takes 1:35 while the second pack takes only 14 seconds total. Building this same image fresh with Docker takes around 2:50 which is in the normal build time range for building a Rails app with Docker. A second build for this particular app with just application-level changes comes to about 35 seconds. The Dockerfile for this app has also been optimized as much as possible with a multi-stage build which also includes caching installed gems and apt packages. Sure, there's always improvements you can make with more caching and more optimized build steps with Docker but the buildpacks give us all of these optimizations out of the box. The bigger win out of this though is that we didn't have to deal with optimizing the build, we were just able to pack it which utilizes all of the buildpack detection support.

Right now with Kamal you have to build an image from your app either locally or with a remote builder with Docker. The idea would be to introduce an approach that builds your app with buildpacks on the backend instead of Docker which also means you don't need a `Dockerfile`.

Maybe something like this:

```
builder:
  local:
    arch: amd64
    builder: heroku/builder:22
```

Or:

```
builder:
  local:
    arch: amd64
    pack: heroku/builder:22
```

Pack seems a little more obvious what it's doing but it also might make sense to have a setting for `builder` that can be set to `docker` or `cnb`(?). Additional customizations to the pack process can be set in a [project.toml](https://buildpacks.io/docs/for-app-developers/how-to/build-inputs/use-project-toml/) file. There's probably a few additional pack-related configuration settings to bubble up but it'd be great to keep as much of that within the pack config, similar to other package managers like bundler, yarn, etc.

Currently the built image with `heroku-22` is a bit bloated but `heroku-24` is being actively worked on and is showing significant improvements to the final image size. 

It's worth noting that this wouldn't replace the default Docker build process but an alternative option that can be simpler to get your app built with. If you want to ensure an incredibly lean and fully customized image you can always reach for Docker but if you're just getting started with Kamal or it's a standardized app you can basically follow the path that Heroku has been busy paving for us. Kamal is built around being able to run a container with Docker but that container doesn't necessarily have to be built with Docker and a Cloud Native Buildpack seems like a nice alternative to introduce. 

Resources:
- Heroku buildpacks repo - https://github.com/heroku/buildpacks
- How to post from @schneems: https://www.schneems.com/2024/05/01/build-a-ruby-on-rails-application-image-in-5-minutes-no-dockerfile-required/
- GitHub discussion post: https://github.com/heroku/buildpacks/discussions/6

**Comment by schneems** [comment]
Thanks for posting the discussion @nickhammond. I wrote the linked article and I'm the buildpack maintainer. I'm happy to answer any CNB or Ruby buildpack-related questions the Kamal core team might have.

> There's probably a few additional pack-related configuration settings to bubble up but it'd be great to keep as much of that within the pack config, similar to other package managers like bundler, yarn, etc.

There's a `project.toml` where most pack-specific application config can live. The idea around pack is that there are good defaults (convention over configuration if you will) and that `pack build` should mostly just work ™️. 

The `heroku/ruby` CNB is still in a "preview" state, so I'm expecting it to get more polished than the current state. It's also worth pointing out that there's more than one game in town. Paketo https://github.com/paketo-buildpacks also has buildpacks, and they've taken a more modular/composable approach.

Also if anyone plays around with buildpacks and has usability feedback, please post on our discussion. @nickhammond's first hand experience has been hugely valuable over at https://github.com/heroku/buildpacks/discussions/6.

**Reply by nickhammond** [reply]
@schneems I started testing out deploying with Kamal via buildpacks by utilizing the `--skip-push` option and packing within the pre-connect/pre-build hook. The idea is that you'd pack the app into an image in those early hooks and then Kamal can just pull down the image on your servers based on the SHA that you're deploying.

```
`pack build nginx-test -t nickhammond/nginx-test:#{ENV['KAMAL_VERSION']}`
`docker push nickhammond/nginx-test:#{ENV['KAMAL_VERSION']}`
```

This almost works but Kamal relies heavily on labels to figure out which containers belong to which service so that it knows it's telling the correct container to stop, start, etc.

Kamal currently does this with the `docker build` command(`--label service="forum"`):

```
docker buildx build --push --builder kamal-nginx-multiarch \
-t nickhammond/nginx-test:df2d2dac98bb9ae5fc15cc48d50b4e3b55a3fe87_uncommitted_8fd8fb4d7663b8c9 \
-t nickhammond/nginx-test:latest-other \
--label service="forum" \
--file Dockerfile .
```

Is it possible to set labels within project.toml? I found this placeholder doc page https://buildpacks.io/docs/for-buildpack-authors/how-to/write-buildpacks/add-labels/ and I'm not seeing anything obvious in the project.toml spec. 

I tried this out with Paketo and it looks like they have an [image-labels](https://github.com/paketo-buildpacks/image-labels) buildpack specifically for this. Is there something similar for Heroku's buildpacks?
```
pack build nginx-test --buildpack paketo-buildpacks/ruby \
  --builder paketobuildpacks/builder-jammy-base \
  --env BP_IMAGE_LABELS="service=other"
```

**Comment by nickhammond** [comment]
Before adding an additional builder to Kamal I wanted to do a simple prototype utilizing the `pre-connect` hook. As a way to demonstrate that we can pack an image the way Heroku would and then utilize Kamal to deploy it. The lingering issue was getting the label added to the image on build for Kamal but there's a paketo buildpack for that. 

Note that a Dockerfile is not necessary, the build-specific artifacts would be the project.toml. This is where you set the buildpacks that your app needs(ruby, node, python, etc), the required label for Kamal, and the files that can be excluded from the final image. 

Here's a rough overview to get an app deployed via buildpacks with Kamal, I'll probably write a longer blog post about it at some point but just wanted to share the core pieces of it. 

_pre-connect_ - This would be what the new builder ends up needing to do.
```
if ENV['KAMAL_COMMAND'] == 'deploy'
  def system!(*args)
    system(*args, exception: true)
  end

  version = ENV['KAMAL_VERSION']
  destination = ENV['KAMAL_DESTINATION']

  image_name = "nickhammond/hey"
  images = {
    pack: "#{image_name}",
    version: "#{image_name}:#{version}"
  }

  build_command = "pack build --platform linux/amd64 #{images[:pack]} -t #{images[:version]}"
  puts "Running: #{build_command}"
  system!(build_command)

  push_version_command = "docker push #{images[:version]}"
  puts "Running: #{push_version_command}"
  system!(push_version_command)
end
```

_config/deploy.yml_ - buildpacks don't expose a port, you have to publish the correct port for your container.
```
publish: 3000
```

_Procfile_ - The default process type is `web`, this boots Puma for us. I haven't tried out adding a worker role but I'd imagine you'd add a process type to Procfile and/or utilize the `cmd` option to boot the correct process.
```
web: bundle exec puma -C config/puma.rb
```

_project.toml_
```
[_]
schema-version = "0.2"
id = "io.buildpacks.hey"
name = "Hey"
version = "1.0.0"

[[io.buildpacks.group]]
uri = "heroku/ruby"

[[io.buildpacks.group]]
uri = "heroku/procfile"

[[io.buildpacks.group]]
uri = "paketo-buildpacks/image-labels"

[[io.buildpacks.build.env]]
name = "BP_IMAGE_LABELS"
value = "service=hey"

[io.buildpacks]
exclude = [
  "/.git/",
  "/.bundle",
  "/config/master.key",
  "/config/credentials/*.key",
  "/.env*"
]
```

I also set my default builder to heroku-24(`pack config default-builder heroku/builder:24`) but this would probably get set as a builder option with Kamal and passed to the pack command as `--builder heroku/builder:24`.

With this just building on `pre-connect` this wouldn't be great for other Kamal commands. Ideally a new builder would be added that does something similar so that it works with all of the Kamal commands and not just deploy. 

Utilizing buildpacks then gives you the familiar Heroku auto-detection, cache, and build process that you're probably familiar with. You would trigger this flow by using the skip push option(`-P`) `kamal deploy -P --verbose` but ideally this would just be configured within your deploy recipe. 

<details>
<summary>pack build --platform linux/amd64 nickhammond/hey -t nickhammond/hey:fbbfc237711ba384fee3c4d40e5b758fae51d292</summary>
<pre><code>
 ===> BUILDING
 [builder] 
 [builder] # Heroku Ruby Buildpack
 [builder] 
 [builder] - Metrics agent
 [builder]   - Skipping install (`barnes` gem not found)
 [builder] - Ruby version `3.2.2` from `Gemfile.lock`
 [builder]   - Using cached Ruby version
 [builder] - Bundler version `2.4.7` from `Gemfile.lock`
 [builder]   - Using cached version
 [builder] - Bundle install
 [builder]   - Loading cached gems
 [builder]   - Skipping `bundle install` (no changes found in /workspace/Gemfile, /workspace/Gemfile.lock, or user configured environment variables)
 [builder]   - ! HELP To force run `bundle install` set `HEROKU_SKIP_BUNDLE_DIGEST=1`
 [builder] - Setting default processes
 [builder]   - Running `bundle list` ... (0.276s)
 [builder]   - Detected rails app (`rails` gem found)
 [builder] - Rake assets install
 [builder]   - Detected rake (`rake` gem found, `Rakefile` found at `/workspace/Rakefile`)
 [builder]   - Running `bundle exec rake -P --trace` .... (1.589s)
 [builder]   - Compiling assets with cache (detected `rake assets:precompile` and `rake assets:clean` via `bundle exec rake -P`)
 [builder]   - Loading cache for /workspace/public/assets
 [builder]   - Loading cache for /workspace/tmp/cache/assets
 [builder]   - Running `bundle exec rake assets:precompile assets:clean --trace`
 [builder] 
 [builder]       ** Invoke assets:precompile (first_time)
 [builder]       ** Invoke assets:environment (first_time)
 [builder]       ** Execute assets:environment
 [builder]       ** Invoke environment (first_time)
 [builder]       ** Execute environment
 [builder]       ** Invoke tailwindcss:build (first_time)
 [builder]       ** Execute tailwindcss:build
 [builder]       Browserslist: caniuse-lite is outdated. Please run:
 [builder]         npx update-browserslist-db@latest
 [builder]         Why you should do it regularly: https://github.com/browserslist/update-db#readme
 [builder]       
 [builder]       Rebuilding...
 [builder]       
 [builder]       Done in 907ms.
 [builder]       ** Execute assets:precompile
 [builder]       ** Invoke assets:clean (first_time)
 [builder]       ** Invoke assets:environment 
 [builder]       ** Execute assets:clean
 [builder]             
 [builder]   - Done (3.248s)
 [builder]   - Storing cache for /workspace/public/assets
 [builder]   - Storing cache for /workspace/tmp/cache/assets
 [builder] - Done (finished in 5.303s)
</code></pre>
</details>

**Comment by nickhammond** [comment]
@schneems PR opened here https://github.com/basecamp/kamal/pull/916. Could use your insight on using secrets for the build process and ensuring those don't end up in the final image. It'd also be great to be able to point to a _.dockerignore_ file for the Exclude list in project.toml, haven't done anything around that yet.
